content/blog/2023-12-11-A DSL for Metastability.md

```diff
@@ -0,0 +1,237 @@
+The pipeline first receives a piece of code written in the DSL, parses it according to a parser that I have written using [Lark](https://lark-parser.readthedocs.io/en/stable/) in python, and then uses the interpreter written in Rust to return a result.
+The execution engine, on the other hand, is currently missing because the borrow-checker does not like me.
+
+# Challenges
+Almost everything in this project has been super challenging, as we have been trying to carve out a DSL from thin air to study metastability.
+As such, I am currently in doubt of almost all of the design choices that I have made thus far.
+The only comforting point is that I have managed to recreate several cases of metastability with the DSL - analyzed with pen and paper - and that these cases are not contrived: they come from real-life experiences.
+Moreover, it seems like the DSL captures most of what is crucial for observing and studying metastability: finite capacity, recurrent cycles of processing, and a concurrent execution setting.
+
+The following list, by no means exhaustive, sheds light on some of the noteworthy challenges:
+
+1. The ownership model of Rust, in tandem with the borrow-checker, is a totally new way of thinking about code for me. Things that are pretty straightforward in other languages become complicated. This problem is exacerbated by the fact that I find it convenient to pack stuff together into a struct, and it is not trivial to me at all how to use a struct's methods on itself with a careful observation of ownership, specially in recursive calls.
+2. As I mentioned earlier, I have decided to make the DSL look like NetKAT. My hunch for this was that the general story in our setting looks very similar to the one studied by NetKAT: requests (packets) travelling around, and getting processed by agents (routers) based on their type (port, switch, etc). However, the semantics are currently drastically different. The important abstraction in NetKAT is a packet, and the semantics revolves around a packet: the network is a mapping of a packet to a set of packets. In our DSL, in contrast, the semantics is much more similar to the classic small-step semantics of, say, IMP, where you have a state (or a store), and you apply some operations on it. My hope was to be able to use something similar to the equational inference system baked into NetKAT, via Kleene algebras, to study metastability. I have not yet seen a clear connection.
+3. On the long run, one would hope to have a performant interpreter, i.e., one that can interpret large systems in a short span of time. The crux of the difficulty lies in parallelizing the execution in the right way. While there are straightforward ways of attempting this, I have not yet found a good way of doing it.
```

Comment:
Indeed; writing parallel interpreters can be really hard. One thing you might consider, in the long term, is to sacrifice parallelism entirely for the straightforward interpreter mode. Instead, you can exploit parallelism at the level of your model checker, i.e., the exploration of a large space of inputs. That exploration can be much more naturally amenable to parallelism.